src/ipc/ipc-zephyr.c

```diff
@@ -79,6 +83,62 @@ static bool message_handler(const struct device *dev, void *arg, uint32_t data,
 	return false;
 }
 
+#ifdef CONFIG_PM_DEVICE
+/**
+ * @brief IPC device suspend handler callback function.
+ * Checks whether device power state should be actually changed.
+ *
+ * @param dev IPC device.
+ * @param arg IPC struct pointer.
+ */
+static int ipc_device_suspend_handler(const struct device *dev, void *arg)
+{
+	struct ipc *ipc = (struct ipc *)arg;
+
+	// we are not entering D3 - return error code bad message
+	if (!(ipc->task_mask & IPC_TASK_POWERDOWN))
+		return -EBADMSG;
+
+	// we are not entering D3 - power state change is not caused by IPC message
+	// from host that would disable the lock.
+	if (pm_policy_state_lock_is_active(PM_STATE_SOFT_OFF, PM_ALL_SUBSTATES))
+		return -EACCES;
+
+	return 0;
+}
+
+/**
+ * @brief IPC device resume handler callback function.
+ * Resets IPC control after context restore.
+ *
+ * @param dev IPC device.
+ * @param arg IPC struct pointer.
+ */
+static int ipc_device_resume_handler(const struct device *dev, void *arg)
+{
+	struct ipc *ipc = (struct ipc *)arg;
+
+	ipc_set_drvdata(ipc, NULL);
+	ipc->task_mask = 0;
+	ipc->pm_prepare_D3 = false;
+
+	// re-enable D3 power state lock
+	pm_policy_state_lock_get(PM_STATE_SOFT_OFF, PM_ALL_SUBSTATES);
+
```

Comment:
This state is is entered by force. That means the all locks are ignored. I don't see why we want to disable lock after pm_state_force.

Reply:
I've verified this once again using debugger - this `pm_policy_state_lock_put(PM_STATE_SOFT_OFF, PM_ALL_SUBSTATES);`
in the `ipc_device_suspend_handler()` function works despite entering power state with `pm_state_force`.
If you comment out line where the lock is unlocked (so the IPC unlocks the lock to allow entering D3):
`// pm_policy_state_lock_put(PM_STATE_SOFT_OFF, PM_ALL_SUBSTATES);`
then, pm_policy_state_lock_is_active() returns `true` value.
In my opinion this lock system should stay as it is a validation for the IPC driver API and the Zephyr PM.
As the IPC message is the only source of signal for the whole system to enter D3 power state,
we should make sure that `ipc->task_mask` is set to `IPC_TASK_POWERDOWN` during driver suspension.

Reply:
The lock over PM_STATE_SOFT_OFF was to be permanent. It has no sense to me that we enter this state by force and remove lock during this.

Reply:
> The lock over PM_STATE_SOFT_OFF was to be permanent. It has no sense to me that we enter this state by force and remove lock during this.

Well, it does to me. Whatever state we force into we need to make sure that internal DSP communication is intact.

Reply:
Fixed, removed the lock that was conceptually wrong and replaced with check whether the IPC queue is empty.